design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,355 @@
+
+#### 2. Validate the Resource Restrictions are defined correctly
+
+In the Azure Authenticator and in the K8s Authenticator we have constraints on 
+how the annotations are defined.
+
+ As can be seen 
+[here](/app/domain/authentication/authn_azure/validate_resource_restrictions.rb), 
+we have the following validations:
+
+  - Validate only permitted restrictions are defined
+    - Here we validate that the `authn-azure` annotations are known. For
+      example, the annotation `authn-azure/subscription-id` is valid, whereas
+      the annotation `authn-azure/unknown` is invalid. We do that to minimize
+      errors from the user, where they will define, for example, the
+      annotation `authn-azure/subscription-ir` (note the typo). In this
+      case we'd rather fail the authentication and point them to their
+      mistake rather than ignore it, and authenticate the request in a weaker
+       granularity than what is expected by the user.
+  - Validate required restrictions exist
+    - For example, in the Azure Authenticator we require hosts to have the
+     `subscription-id` and the `resource-group` annotations. In the K8s
+      Authenticator we require the `namespace` annotation.
+  - Validate mutually exclusive restrictions
+    - For example, in the Azure Authenticator we allow hosts to have
+     only one of the `user-assigned-identity` and the `system-assigned-identity`
+     annotations. In the K8s we allow hosts to have only one of
+      the annotations `deployment`, `deployment-config` & `stateful-set`.
+
+Also here we can see some generalization.
+
+These need to be defined for each authenticator.
+To simplify that, We can define a simple class to aggregate this information.
+It will allow for each restriction to define its behavior at once, and prevent
+string duplication across the code.
+
+The class will be in the format:
+
+```ruby
+class ResourceRestrictionsConstraints
+
+  attr_reader :permitted, :required, :mutually_exclusive
+
+  def initialize
+    @permitted = []
+    @required = []
+    @mutually_exclusive = {}
+  end
+
+  def add(restriction_name:, required: false, exclusive_group_name: nil)
+    @permitted << restriction_name
+    @required << restriction_name if required
+    (@mutually_exclusive[exclusive_group_name] ||= []) << restriction_name if exclusive_group_name.present?
+    self
+  end
+end
+```
+
+It will be used by each authenticator to define how to validate the configured
+restrictions. These are the ones that were defined in the policy for the host,
+regardless of the request.
+
+Example how it will look like for `authn-azure`:
+
+```ruby
+@constraints = ResourceRestrictionsConstraints.new
+  .add(restriction_name: "subscription-id", 		 required: true)
+  .add(restriction_name: "resource-group", 			 required: true)
+  .add(restriction_name: "user-assigned-identity",   exclusive_group_name: "identity")
+  .add(restriction_name: "system-assigned-identity", exclusive_group_name: "identity")
```

Comment:
is there a case where `exclusive_group_name` would be different than "identity"? In other words would there be more than one `exclusive_group_name` in our current usecase? Or are you predicting a future scenario where there will be one?

Reply:
That's a great question!
While I wrote this class, it bothered me a lot to generalize it to support a single exclusive group.
There's no reason to limit it in such a way.
Still,  the effort to support multiple groups is worth it so the API won't change once this is needed.
Also, this is a very small change, and it makes the code more readable, IMO.